docs/dapp/sapphire/precompiles.md

```diff
@@ -60,14 +60,33 @@ Feel free to discover other convenient libraries for Solidity inside the
 * Gas cost: 10,000 minimum plus 240 per output word plus 60 per word of
   the personalization string.
 
-Generate `num_bytes` pseudo-random bytes, with an optional
-personalization string added into the hashing algorithm to increase
-domain separation when needed.
+Generate `num_bytes` pseudo-random bytes, with an optional personalization
+string (`pers`) added into the hashing algorithm to increase domain separation
+when needed.
 
 ```solidity
 bytes memory randomPad = Sapphire.randomBytes(64, "");
 ```
 
+:::danger Prior to 0.5.3
+All view queries and simulated transactions (via `eth_call`) would receive the
+same entropy in-between blocks if they use the same `num_bytes` and `pers` parameters.
+
+This does not affect transactions, so if your contract requires confidentiality please
+generate a secret in the constructor and use it from view calls.
+
+```solidity
+Sapphire.randomBytes(64, abi.encodePacked(msg.sender, this.perContactSecret));
+```
+:::
+
+The mode (e.g. simulation or 'view call' vs transaction execution) is fed to TupleHash (among other
+block-dependent components) to derive the "key id", which is then used to derive a per-block VRF key
+from epoch-ephemeral entropy (using KMAC256 and cSHAKE) so a different "key id" will result in a
+unique per-block VRF key. This per-block VRF key is then used to create the per-block root RNG which
+is then used to derive domain-separated (using Merlin transcripts) per-transaction random RNGs which
+are then exposed via this precompile.
```

Comment:
Maybe move this under its own `### Implementation Details` or similar. Would also be nice if you could find links for KMAC256, cSHAKE, Merlin and TupleHash.